mathics/builtin/makeboxes.py

```diff
@@ -65,94 +71,119 @@ def make_boxes_infix(
                 s = repr(value)
             else:
-                with mpmath.workprec(expr.get_precision()):
-                    value = expr.to_mpmath()
-                    s = mpmath.nstr(value, dps(expr.get_precision()) + 1)
+                with mpmath.workprec(real.get_precision()):
+                    value = real.to_mpmath()
+                    s = mpmath.nstr(value, dps(real.get_precision()) + 1)
         else:
-            with mpmath.workprec(expr.get_precision()):
-                value = expr.to_mpmath()
-                s = mpmath.nstr(value, n)
+            with mpmath.workprec(real.get_precision()):
+                value = real.to_mpmath()
+                s = mpmath.nstr(value, digits)
 
-        # sign prefix
+        # Set sign prefix.
         if s[0] == "-":
             assert value < 0
-            nonnegative = 0
+            is_nonnegative = False
             s = s[1:]
         else:
             assert value >= 0
-            nonnegative = 1
-        # exponent (exp is actual, pexp is printed)
+            is_nonnegative = True
+        # Set exponent. ``exponent`` is actual, ``pexp`` of ``number_form()`` is printed.
         if "e" in s:
-            s, exp = s.split("e")
-            exp = int(exp)
+            s, exponent = s.split("e")
+            exponent = int(exponent)
             if len(s) > 1 and s[1] == ".":
                 # str(float) doesn't always include '.' if 'e' is present.
                 s = s[0] + s[2:].rstrip("0")
         else:
-            exp = s.index(".") - 1
-            s = s[: exp + 1] + s[exp + 2 :].rstrip("0")
+            exponent = s.index(".") - 1
+            s = s[: exponent + 1] + s[exponent + 2 :].rstrip("0")
 
-            # consume leading '0's.
+            # Normalize exponent: remove leading '0's after the decimal point
+            # and adjust the exponent accordingly.
             i = 0
-            while s[i] == "0":
+            while i < len(s) and s[i] == "0":
                 i += 1
-                exp -= 1
+                exponent -= 1
             s = s[i:]
 
-        # add trailing zeros for precision reals
-        if n is not None and not expr.is_machine_precision() and len(s) < n:
-            s = s + "0" * (n - len(s))
-    return s, exp, nonnegative
+        # Add trailing zeros for precision reals.
+        if digits is not None and not real.is_machine_precision() and len(s) < digits:
+            s = s + "0" * (digits - len(s))
+    return s, exponent, is_nonnegative
 
 
-def number_form(expr, n, f, evaluation: Evaluation, options: dict):
+# FIXME the return type should be a NumberForm, not a String.
+def number_form(
+    value: Union[Real, Integer],
+    digits: Optional[int],
+    digits_after_decimal_point: Optional[int],
+    evaluation: Evaluation,
+    options: dict,
+) -> String:
     """
-    Converts a Real or Integer instance to Boxes.
+    Converts a Real or Integer value to a String. FIXME: the return should be a Numberform.
```

Comment:
Numberform is a Python type? As far as I can see, the function returns `String` objects.

Reply:
I am not sure I understand. NumberForm _currently_ returns a String in Mathics3, but should it? I don't that is what it does in WMA. 

Here is what I see in wolframscript: 

```
In[1]:= NumberForm[5, {5, 3}]                                                   

Out[1]//NumberForm= 5  (* Note this is tagged on output as NumberForm, and that suggests the result is not a String. In Mathics3 the result form is None.*)

In[2]:= Head[NumberForm[5, {5, 3}]]                                             

Out[2]= NumberForm  (* This suggests the result is a NumberForm. Head [] is not does not have  A_HOLD_ attributes *)

In[3]:= Head[ToString[NumberForm[5, {5, 3}]]                                    ]                                                                               

Out[3]= String  (* Not surprising; but just verifying. *)                                                
```

Suppose I write: Something like 

```
FIXME: the return value should be a NumberForm Value instead of a String. 
```

Is this correct? Or suggest what would be more correct. Thanks!

Reply:
OK, maybe what is misleading is the name of the function.  The return value is the result of formatting a NumberForm[...] as a String. Maybe "numberform_to_string" would be a more enlightening name...